File: backend/routers/assistant.py
"""Structured, grounded streaming assistant API."""

import json
import logging
from typing import Any, Literal, Optional

from fastapi import APIRouter
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from routers.monitoring import get_monitoring_snapshot
from services.chat_service import stream_chat
from services.decision_assistant import build_grounding

router = APIRouter()
logger = logging.getLogger(__name__)


class AssistantMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str


class AssistantRequest(BaseModel):
    message: str = Field(min_length=1, max_length=2000)
    history: list[AssistantMessage] = Field(default_factory=list)
    context: dict[str, Any] = Field(default_factory=dict)
    file_id: Optional[int] = None


def sse(event_type: str, data: dict[str, Any]) -> str:
    return f"data: {json.dumps({'type': event_type, **data}, ensure_ascii=False)}\n\n"


@router.post("/stream", summary="基于监测上下文的流式研判")
async def assistant_stream(payload: AssistantRequest):
    async def event_stream():
        try:
            snapshot = await get_monitoring_snapshot(payload.file_id)
            grounding = build_grounding(payload.message, snapshot)
            provenance = snapshot["provenance"]
            yield sse(
                "context_ready",
                {
                    "data_version": provenance["data_version"],
                    "model_version": provenance["model_version"],
                    "fallback_mode": provenance["mode"],
                    "missing_fields": provenance["missing_fields"],
                },
            )
            yield sse("tool_started", {"label": "读取监测快照"})
            yield sse("tool_finished", {"label": "读取监测快照"})

            prompt = (
                f"用户问题：{payload.message}\n"
                f"页面筛选上下文：{json.dumps(payload.context, ensure_ascii=False)}\n"
                f"可引用的监测数据：\n{grounding['context_text']}"
            )
            messages = [item.model_dump() for item in payload.history[-10:]]
            messages.append({"role": "user", "content": prompt})
            buffered_tokens: list[str] = []
            model_failed = False
            try:
                async for token in stream_chat(
                    messages,
                    temperature=0.2,
                    max_tokens=900,
                ):
                    if "[错误]" in token:
                        model_failed = True
                        break
                    buffered_tokens.append(token)
            except Exception:
                logger.warning("模型不可用，切换规则研判", exc_info=True)
                model_failed = True

            warnings: list[str] = []
            if model_failed or not buffered_tokens:
                warnings.append("MODEL_UNAVAILABLE_RULE_FALLBACK")
                yield sse(
                    "answer_delta",
                    {"content": grounding["fallback_answer"]},
                )
            else:
                for token in buffered_tokens:
                    yield sse("answer_delta", {"content": token})

            yield sse(
                "citations_ready",
                {"citations": grounding["citations"]},
            )
            yield sse("actions_ready", {"actions": grounding["actions"]})
            yield sse("completed", {"warnings": warnings})
        except Exception:
            logger.exception("结构化研判失败")
            yield sse("error", {"content": "研判服务暂不可用，请稍后重试"})

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )

File: frontend/src/store/assistant.js
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { streamAssistant } from '../api/assistant'
import { consumeSseChunk } from '../features/assistant/sse'

const warningLabels = {
  MODEL_UNAVAILABLE_RULE_FALLBACK: '模型服务暂不可用，当前回答由可追溯规则生成',
}

export const useAssistantStore = defineStore('assistant', () => {
  const messages = ref([])
  const progress = ref('')
  const citations = ref([])
  const actions = ref([])
  const warnings = ref([])
  const isStreaming = ref(false)
  const metadata = ref({})
  let controller = null

  function applyEvent(event, assistantMessage) {
    if (event.type === 'context_ready') metadata.value = event
    if (event.type === 'tool_started') progress.value = `正在${event.label}`
    if (event.type === 'tool_finished') progress.value = `${event.label}完成`
    if (event.type === 'answer_delta') assistantMessage.content += event.content || ''
    if (event.type === 'citations_ready') citations.value = event.citations || []
    if (event.type === 'actions_ready') actions.value = event.actions || []
    if (event.type === 'completed') {
      warnings.value = (event.warnings || []).map((item) => warningLabels[item] || item)
      progress.value = '分析完成'
    }
    if (event.type === 'error') {
      warnings.value = [event.content || '研判服务暂不可用']
      progress.value = '分析中断'
    }
  }

  async function send(message, context, history = []) {
    if (!message.trim() || isStreaming.value) return
    const priorHistory = history.length
      ? history
      : messages.value.map(({ role, content }) => ({ role, content }))
    const userMessage = { role: 'user', content: message.trim() }
    const assistantMessage = { role: 'assistant', content: '' }
    messages.value.push(userMessage, assistantMessage)
    citations.value = []
    actions.value = []
    warnings.value = []
    progress.value = '正在获取上下文'
    isStreaming.value = true
    controller = new AbortController()

    try {
      const response = await streamAssistant({
        message: userMessage.content,
        history: priorHistory,
        context,
        file_id: context.fileId || null,
      }, controller.signal)
      if (!response.ok) throw new Error(`HTTP ${response.status}`)

      const reader = response.body.getReader()
      const decoder = new TextDecoder()
      let buffer = ''
      while (true) {
        const { done, value } = await reader.read()
        if (done) break
        const parsed = consumeSseChunk(buffer, decoder.decode(value, { stream: true }))
        buffer = parsed.remainder
        parsed.events.forEach((event) => applyEvent(event, assistantMessage))
      }
    } catch (error) {
      if (error.name !== 'AbortError') {
        warnings.value = ['研判服务暂不可用，请重试']
        progress.value = '分析中断'
      }
    } finally {
      isStreaming.value = false
      controller = null
    }
  }

  function cancel() {
    controller?.abort()
  }

  function reset() {
    cancel()
    messages.value = []
    progress.value = ''
    citations.value = []
    actions.value = []
    warnings.value = []
  }

  return {
    messages,
    progress,
    citations,
    actions,
    warnings,
    isStreaming,
    metadata,
    send,
    cancel,
    reset,
  }
})
